### DISCUSSION TITLE
[eslint-mdx] focus on next version of mdx

### DISCUSSION THREAD

**Discussion by JounQin** [mdx-js/mdx#1484] [General] [answered]
@wooorm Any advise on how should I start?

All I can find is https://github.com/mdx-js/mdx/pull/1382#issuecomment-745985440

But I try to run that codes again, there are errors:

```log
/Users/JounQin/Workspaces/GitHub/eslint-mdx/node_modules/micromark-extension-mdx-jsx/lib/factory-tag.js:701
    throw new VMessage(
    ^
[5:6: Unexpected character `/` (U+002F) after self-closing slash, expected `>` to end the tag (note: JS comments in JSX tags are not supported in MDX)] {
  reason: 'Unexpected character `/` (U+002F) after self-closing slash, expected `>` to end the tag (note: JS comments in JSX tags are not supported in MDX)',
  line: 5,
  column: 6,
  location: {
    start: { line: 5, column: 6, offset: 42, _index: 7, _bufferIndex: 5 },
    end: { line: null, column: null }
  },
  source: 'micromark-extension-mdx-jsx',
  ruleId: 'unexpected-character'
}
```

I hope to have an easy-to-use API to get estree AST with comments.

**Comment by JounQin** [comment, marked as answer]
Additional, [remark-mdx](https://github.com/mdx-js/mdx/tree/main/packages/remark-mdx)'s codes are not synced to npm.

I tried to copy it locally and use it with:

```js
const unified = require('unified')
const parse = require('remark-parse')

const mdx = require('./remark-mdx')

const doc = `
<>
  {/* MDX is awesome! */}
  <div
    // MDX is awesome!
    style={{ color: "white", backgroundColor: "black", padding: "24px 32px" }}
  ></div>
</>;
`

const processor = unified().use(parse).use(mdx).freeze()

const root = processor.parse(doc)

console.log(root)
```

The error is the same as above:

```log
/Users/JounQin/Workspaces/GitHub/eslint-mdx/node_modules/micromark-extension-mdx-jsx/lib/factory-tag.js:701
    throw new VMessage(
    ^
[5:6: Unexpected character `/` (U+002F) after self-closing slash, expected `>` to end the tag (note: JS comments in JSX tags are not supported in MDX)] {
  reason: 'Unexpected character `/` (U+002F) after self-closing slash, expected `>` to end the tag (note: JS comments in JSX tags are not supported in MDX)',
  line: 5,
  column: 6,
  location: {
    start: { line: 5, column: 6, offset: 42, _index: 7, _bufferIndex: 5 },
    end: { line: null, column: null }
  },
  source: 'micromark-extension-mdx-jsx',
  ruleId: 'unexpected-character'
}
```

**Reply by JounQin** [reply]
OK, I tried a simpler conent and it's working now, so it should be a bug instead. @wooorm 

```mdx
<>
  {/* MDX is awesome! */}
  <div
    style={{ color: "white", backgroundColor: "black", padding: "24px 32px" }}
  ></div>
</>
```

The trailing `;` and comment `// MDX is awesome!` breaks the parser.

**Reply by wooorm** [reply]
I don’t see a semicolon? Can you make this reproducible?

**Reply by JounQin** [reply]
https://github.com/mdx-js/mdx/discussions/1484#discussioncomment-480715

**Reply by wooorm** [reply]
Read the error message: '(note: JS comments in JSX tags are not supported in MDX)'

I never see folks putting comments inside tags. How common is that?

**Reply by JounQin** [reply]
Well, I don't know, but it is valid jsx.

**Reply by wooorm** [reply]
Yes, and it’s documented: https://github.com/micromark/mdx-state-machine#73-deviations-from-jsx. Plus it’s explained in the error message. As it’s (essentially) never used, similar to some other JSX features, I didn’t implement it.

**Reply by JounQin** [reply]
OK, fair enough for this case, then what about the `semi`?

```jsx
import { Basic } from './basic'

<Basic>
# abc
</Basic>;
```

```log
/Users/JounQin/Workspaces/GitHub/eslint-mdx/test/fixtures/jsx.mdx
  0:0  error  Parsing error: Cannot close `mdxJsxTextTag` (5:1-5:9): a different token (`paragraph`, 5:1-5:10) is open
```

**Reply by wooorm** [reply]
Semicolons don’t exist in MDX. I think we’ve talked about this before?

**Reply by JounQin** [reply]
https://github.com/mdx-js/eslint-mdx/issues/183#issuecomment-618903311

And it does work in v1.

![image](https://user-images.githubusercontent.com/8336744/111300026-9c8d1580-868b-11eb-802a-27acb8f99d34.png)

**Reply by wooorm** [reply]
That is a different example?

**Reply by JounQin** [reply]
> That is a different example?

How about this one

![image](https://user-images.githubusercontent.com/8336744/111301156-d3aff680-868c-11eb-966a-e214aaec7105.png)

**Reply by wooorm** [reply]
First off: the semicolon shouldn‘t be there. It doesn’t make sense. MDX !== JS. You can’t put a `console.log` in there either. So what’s the point of this discussion?

For why it works this way, and differently now than in MDX 1, see: https://github.com/micromark/mdx-state-machine#74-common-mdx-gotchas and the surrounding sections!

**Reply by JounQin** [reply]
I thought that you meant we should treat `;` as text at https://github.com/mdx-js/eslint-mdx/issues/183#issuecomment-618922977 😂

**Reply by wooorm** [reply]
ah, there it was! OK, that sounds a bit different!

My reasoning there was: it *should not be part of JSX*. Because MDX does not have semicolons. If you put it in MDX, it’s going to “show up”, not be hidden.

What I am arguing here is a bit different. In markdown, you have blocks, right? And inline.

````markdown
```js
// This is a code block.
```

This is `inline`.
````

Blocks can be blocks, inline is inline, but they can’t **mix**:

````markdown
```js
// this next semicolon doesn’t work in markdown! The code isn’t closed by these ticks + semicolon
```;
````

So we start with markdown, and then we add JSX to get MDX. We also want to allow blocks and inline. But how to differentiate between them?

So, when parsing your example:

```mdx
<Basic>
# abc
</Basic>;
```

The parser first sees `<Basic>` on its own line: well, that’s a “block”, right? Makes sense to me.
But `</Basic>;` can’t be a block. Similar to `whatever here </Basic> and other stuff here`. That looks like “inline”?

So we have a problem: one is “block”, the other is “inline”. And that crashes.

Don’t put the semicolon there!

**Reply by JounQin** [reply]
I'm fine with this limitation, would be great to document this at https://github.com/micromark/mdx-state-machine#74-common-mdx-gotchas.
